srtcore/platform_sys.h

```diff
@@ -71,6 +71,11 @@
    #include <unistd.h>
 #endif
 
+#if defined(__clang__)
+#pragma clang diagnostic ignored "-Wshorten-64-to-32"
+#pragma clang diagnostic ignored "-Wambiguous-reversed-operator"
+#endif
```

Comment:
Could you please paste, somewhere in comments maybe, where exactly these warnings are being reported?

Reply:
```
In file included from /Users/oliver/Git/source/ovlibs/ovnet/thirdparty/srt/srtcore/window.cpp:58:
In file included from thirdparty/srt/srtcore/window.h:61:
thirdparty/srt/srtcore/packet.h:157:54: error: implicit conversion loses integer precision: 'size_t' (aka 'unsigned long') to 'uint32_t' (aka 'unsigned int') [-Werror,-Wshorten-64-to-32]
    return SEQNO_CONTROL::mask | SEQNO_MSGTYPE::wrap(size_t(type));
                                 ~~~~~~~~~~~~~       ^~~~~~~~~~~~
/Users/oliver/Git/source/ovlibs/ovnet/thirdparty/srt/srtcore/window.cpp:96:29: error: implicit conversion loses integer precision: 'long long' to 'const int' [-Werror,-Wshorten-64-to-32]
            const int rtt = count_microseconds(currtime - r_aSeq[i].tsTimeStamp);
                      ~~~   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/oliver/Git/source/ovlibs/ovnet/thirdparty/srt/srtcore/window.cpp:115:38: error: implicit conversion loses integer precision: 'unsigned long' to 'int' [-Werror,-Wshorten-64-to-32]
   for (int j = r_iTail, n = r_iHead + size; j < n; ++ j)
                         ~   ~~~~~~~~^~~~~~
/Users/oliver/Git/source/ovlibs/ovnet/thirdparty/srt/srtcore/window.cpp:125:26: error: implicit conversion loses integer precision: 'long long' to 'const int' [-Werror,-Wshorten-64-to-32]
         const int rtt = count_microseconds(currtime - r_aSeq[j].tsTimeStamp);
                   ~~~   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/oliver/Git/source/ovlibs/ovnet/thirdparty/srt/srtcore/window.cpp:179:24: error: implicit conversion loses integer precision: 'size_t' (aka 'unsigned long') to 'int' [-Werror,-Wshorten-64-to-32]
   for (int i = 0, n = asize; i < n; ++ i)
                   ~   ^~~~~
/Users/oliver/Git/source/ovlibs/ovnet/thirdparty/srt/srtcore/window.cpp:195:17: error: implicit conversion loses integer precision: 'unsigned long' to 'int' [-Werror,-Wshorten-64-to-32]
      bytesps = (unsigned long)ceil(1000000.0 / (double(sum) / double(bytes)));
              ~ ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/oliver/Git/source/ovlibs/ovnet/thirdparty/srt/srtcore/window.cpp:243:24: error: implicit conversion loses integer precision: 'size_t' (aka 'unsigned long') to 'int' [-Werror,-Wshorten-64-to-32]
   for (int i = 0, n = psize; i < n; ++ i)
                   ~   ^~~~~
7 errors generated.
```

```
In file included from /Users/oliver/Git/source/ovlibs/ovnet/thirdparty/srt/srtcore/core.cpp:66:
In file included from thirdparty/srt/srtcore/api.h:63:
thirdparty/srt/srtcore/cache.h:116:20: error: ISO C++20 considers use of overloaded operator '==' (with operand types 'srt::CInfoBlock' and 'srt::CInfoBlock') to be ambiguous despite there being a unique best viable function [-Werror,-Wambiguous-reversed-operator]
         if (*data == ***i)
             ~~~~~ ^  ~~~~
/Users/oliver/Git/source/ovlibs/ovnet/thirdparty/srt/srtcore/core.cpp:4606:19: note: in instantiation of member function 'srt::CCache<srt::CInfoBlock>::lookup' requested here
    if (m_pCache->lookup(&ib) >= 0)
                  ^
In file included from /Users/oliver/Git/source/ovlibs/ovnet/thirdparty/srt/srtcore/core.cpp:66:
In file included from thirdparty/srt/srtcore/api.h:63:
thirdparty/srt/srtcore/cache.h:256:9: note: ambiguity is between a regular call to this operator and a call with the argument order reversed
   bool operator==(const CInfoBlock& obj);
        ^
thirdparty/srt/srtcore/cache.h:146:20: error: ISO C++20 considers use of overloaded operator '==' (with operand types 'srt::CInfoBlock' and 'srt::CInfoBlock') to be ambiguous despite there being a unique best viable function [-Werror,-Wambiguous-reversed-operator]
         if (*data == ***i)
             ~~~~~ ^  ~~~~
/Users/oliver/Git/source/ovlibs/ovnet/thirdparty/srt/srtcore/core.cpp:6221:19: note: in instantiation of member function 'srt::CCache<srt::CInfoBlock>::update' requested here
        m_pCache->update(&ib);
                  ^
In file included from /Users/oliver/Git/source/ovlibs/ovnet/thirdparty/srt/srtcore/core.cpp:66:
In file included from thirdparty/srt/srtcore/api.h:63:
thirdparty/srt/srtcore/cache.h:256:9: note: ambiguity is between a regular call to this operator and a call with the argument order reversed
   bool operator==(const CInfoBlock& obj);
        ^
thirdparty/srt/srtcore/cache.h:179:28: error: ISO C++20 considers use of overloaded operator '==' (with operand types 'srt::CInfoBlock' and 'srt::CInfoBlock') to be ambiguous despite there being a unique best viable function [-Werror,-Wambiguous-reversed-operator]
            if (*last_data == ***i)
                ~~~~~~~~~~ ^  ~~~~
thirdparty/srt/srtcore/cache.h:256:9: note: ambiguity is between a regular call to this operator and a call with the argument order reversed
   bool operator==(const CInfoBlock& obj);
        ^
3 errors generated.
```

Reply:
I think breaking the gauge that shows a problem is not the best solution of the problem.

1. Implicit conversion of `type` to `size_t` to be next used as `uint32_t` should be fixed by converting it first to `uint32_t`.
2. The value of rtt if assigned from `int64_t` should be of type `int64_t`, not int. Further passing it to any calculations relying on int should then use an explicit conversion.
3. Likely `j` in the loop, unless there's a risk of a negative value, should also use the `size_t` type.
4. `bytesps` gets assigned a value that is explicitly converted - should then use the exact type of `bytesps`.
5. The overloaded operator== for `CInfoBlock` likely lacks `const`.

Would you be able to try these above and retest if this also fixes the problems? If not, would you be able to send the list of the remaining problems so that I can prepare another fix?

Reply:
1. Ok.
2. Maybe better to just cast to int straightaway from count_microseconds(currtime - r_aSeq[j].tsTimeStamp) as the precedent in SRT seems to be that RTT is an int.
3. If you do this, you get warnings further down, for example r_iTail = (i + 1) % size now complains because r_iTail is an int, but the calculation is now size_t. Basically there is a mismatch throughout SRT where sizes are sometimes size_t, but the rest of the time int. Sometimes you set int to negative numbers/return (-1) to indicate an invalid value, so size_t can't be used everywhere. You could use ssize_t but that's not supported everywhere, so....
4. and 5. yes, trivial changes. I've done 5 already.

Reply:
Yes, I get that, there are many though, not just the ones above.

I question the use of size_t for sizes at all when the core API appears to rely on int.

Take this snippet, for example:

```
    if (m_iMaxPayloadSize == -1)
    {
        int plsize = data.ps->core().OPT_PayloadSize();
        HLOGC(gmlog.Debug,
              log << "CUDTGroup::add: taking MAX payload size from socket @" << data.ps->m_SocketID << ": " << plsize
                  << " " << (plsize ? "(explicit)" : "(unspecified = fallback to 1456)"));
        if (plsize == 0)
            plsize = SRT_LIVE_MAX_PLSIZE;
        // It is stated that the payload size
        // is taken from first, and every next one
        // will get the same.
        m_iMaxPayloadSize = plsize;
    }
```

OPT_PayloadSize is a size_t so int plsize = data.ps->core().OPT_PayloadSize();

Sure, we can cast the plsize to an int, because clearly we're operating in the int range for SRT_LIVE_MAX_PLSIZE, and m_iMaxPayloadSize. But why is OPT_PayloadSize() a size_t in the first place, if the internals of SRT rely on int?

Reply:
Yes, that should be definitely refaxed, but I'd have to investigate further why these values get two different types. Shortening 64 to 32 should be then likely silenced, as less troublesome, at least for the time being.

Reply:
Ok, I've worked through the remaining clang ones and casted in the way that makes sense to me.

Just going to re-enable the Windows warnings I had disabled, and fix any remaining ones on that platform.

Reply:
Just working through the remaining Windows ones.

This code:

```
    // Note: Intentionally the "fixed char size" types are used despite using
    // character size dependent FormatMessage (instead of FormatMessageA) so that
    // your compilation fails when you use wide characters.
    // The problem is that when TCHAR != char, then the buffer written this way
    // would have to be converted to ASCII, not just copied by strncpy.
    FormatMessage(0
            | FORMAT_MESSAGE_ALLOCATE_BUFFER
            | FORMAT_MESSAGE_FROM_SYSTEM
            | FORMAT_MESSAGE_IGNORE_INSERTS,
            NULL, // no lpSource
            errnum, // dwMessageId (as controlled by FORMAT_MESSAGE_FROM_SYSTEM)
            MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
            // This below parameter normally should contain a pointer to an allocated buffer,
            // and this way it's LPTSTR. But when FORMAT_MESSAGE_ALLOCATE_BUFFER, then it is
            // expected to be a the value of LPTSTR* type, converted to LPTSTR, that designates
            // a pointer to a variable of type LPTSTR, to which the newly allocated buffer is
            // assigned. This buffer should be freed afterwards using LocalFree().
            (LPSTR)&lpMsgBuf,
            0, NULL);

    if (lpMsgBuf)
    {
        strncpy(buf, lpMsgBuf, buflen-1);
        buf[buflen-1] = 0;
        LocalFree((HLOCAL)lpMsgBuf);
    }
    else
    {
        SysStrError_Fallback(errnum, buf, buflen);
    }
```

Gives the warning incompatible types - from 'LPSTR' to 'LPWSTR'

Looking at the comment above - it seems to suggest we can't use SRT when compiled with wide characters. Thing is, SRT works just fine in my experience (many years) when compiled as wide char, so what would be the preferred workaround to deal with this warning?

Reply:
Looks like so. Especially that you skipped this line:

Reply:
What I mean is, SysStrError is expected to return a char* to the caller (to write into log macros, etc) so why does this function need to anything with wide chars at all, it can just use FormatMessageA?

Reply:
Well, as long as this form is perfectly portable (that is, Windows doesn't recommend to use FormatMessageA directly), that would be a better solution. @jlsantiago0 what do you think?